### PR TITLE
python310Packages.twisted: adopt, run tests

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by risicle** [comment]
See also #168407

**Comment by risicle** [comment]
On macos 10.15:

```
[ERROR]
Traceback (most recent call last):
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/trial/runner.py", line 691, in loadByName
    return self.suiteFactory([self.findByName(name, recurse=recurse)])
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/trial/runner.py", line 451, in findByName
    obj = reflect.namedModule(searchName)
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/python/reflect.py", line 156, in namedModule
    topLevel = __import__(name)
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/internet/test/test_process.py", line 1091, in <module>
    PTYProcessTestsBuilder_EPollReactorTests.test_openFileDescriptors.skip = "invalid syntax"
builtins.NameError: name 'PTYProcessTestsBuilder_EPollReactorTests' is not defined

twisted.internet.test.test_process
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/trial/runner.py", line 691, in loadByName
    return self.suiteFactory([self.findByName(name, recurse=recurse)])
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/trial/runner.py", line 451, in findByName
    obj = reflect.namedModule(searchName)
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/python/reflect.py", line 156, in namedModule
    topLevel = __import__(name)
  File "/nix/store/sjviiy408rwiksarcjgivhqx8d94iky1-python3.9-twisted-22.4.0/lib/python3.9/site-packages/twisted/internet/test/test_unix.py", line 1018, in <module>
    UNIXTestsBuilder_EPollReactorTests.test_sendFileDescriptorTriggersPauseProducing.skip = "sendFileDescriptor producer was not paused"
builtins.NameError: name 'UNIXTestsBuilder_EPollReactorTests' is not defined

twisted.internet.test.test_unix
-------------------------------------------------------------------------------
Ran 11156 tests in 85.458s

FAILED (skips=1633, errors=2, successes=9521)
```

**Comment by SuperSandro2000** [comment]
Waiting for #171354, a merge from master to staging and then going to rebase this PR.
---
@ofborg eval

File: pkgs/development/python-modules/twisted/default.nix [hunk 1]
```diff
@@ -1,54 +1,153 @@
...
   checkPhase = ''
-    ${python.interpreter} -m unittest discover -s src/twisted/test
+    export SOURCE_DATE_EPOCH=315532800
+    export PATH=$out/bin:$PATH
+    ${python.interpreter} -m twisted.trial -j$NIX_BUILD_CORES twisted
   '';
-  # Tests require network
-  doCheck = false;
+
+  passthru.tests = {
```
Comment (exarkun) [inline pkgs/development/python-modules/twisted/default.nix:137]: What's all this?

File: pkgs/tools/networking/p2p/tahoe-lafs/default.nix [hunk 2]
```diff
@@ -60,7 +60,8 @@ python3Packages.buildPythonApplication rec {
     html5lib magic-wormhole netifaces pyasn1 pycrypto pyutil pyyaml recommonmark
     service-identity simplejson sphinx_rtd_theme testtools treq twisted zfec
     zope_interface
-  ];
+  ] ++ twisted.extras-require.tls
+    ++ twisted.extras-require.conch;
```
Comment (exarkun) [inline pkgs/tools/networking/p2p/tahoe-lafs/default.nix:64]: fwiw I have been trying to adopt this package but I can't get a review on https://github.com/NixOS/nixpkgs/pull/166601/files

File: pkgs/development/python-modules/twisted/default.nix (hunk 1 above)
Comment (risicle) [reply pkgs/development/python-modules/twisted/default.nix]: @exarkun see #168407 for an explanation.

It's still a slightly experimental approach to testing heavily depended packages but if we find it causes problems we can always abandon the idea.

**Pull request merged**